lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt

```diff
@@ -548,26 +548,52 @@ internal class EvaluatingCompiler(
                         }
                     }
                     TypingMode.PERMISSIVE -> {
+                        // if null, then StaticTypes contains one or more Int Type.
+                        val nonIntType = staticTypes.filterNot {
+                            it is IntType
+                        }.ifEmpty { null }
                         val biggestIntegerType = staticTypes.filterIsInstance<IntType>().maxBy {
                             it.rangeConstraint.numBytes
                         }
-                        when (biggestIntegerType) {
-                            is IntType -> {
-                                val validator = integerValueValidator(biggestIntegerType.rangeConstraint.validRange)
-
+                        val validateResult = { validator: ((ExprValue) -> Boolean)?, naryResult: ExprValue ->
+                            errorSignaler.errorIf(
+                                !validator?.let { it(naryResult) }!!,
+                                ErrorCode.EVALUATOR_INTEGER_OVERFLOW,
+                                { ErrorDetails(metas, "Integer overflow", errorContextFrom(metas)) },
+                                { naryResult }
+                            )
+                        }
+                        when (nonIntType) {
+                            // only intType exists
+                            null -> {
+                                // biggestIntegerType can not be null in this case
+                                val validator = integerValueValidator(biggestIntegerType!!.rangeConstraint.validRange)
                                 thunkFactory.thunkEnv(metas) { env ->
                                     val naryResult = computeThunk(env)
-                                    errorSignaler.errorIf(
```

Comment:
I feel like the fix can be much simpler than what's currently in this PR. If I understand the original issue, this `errorSignaler` is giving an error when the resulting evaluation type is a non-Int (e.g. decimal) when the `validator` expects an Int. A simple fix should be to case on when the resulting `naryResult`'s type is an Int:

```Kotlin
when (naryResult.type) {
    ExprValueType.INT -> errorSignaler.errorIf(
        !validator(naryResult),
        ErrorCode.EVALUATOR_INTEGER_OVERFLOW,
        { ErrorDetails(metas, "Integer overflow", errorContextFrom(metas)) },
        { naryResult }
    )
    else -> naryResult
}
```

Reply:
If I understand you statement correctly, you are suggesting that the logic should be somewhat like below:
1. If no `IntType` in StaticType, return the computeThunk directly.
2. If there is at least one `IntType` in StaticType, then we first check if the `naryResult` is Int, if so, we run the validator, if not we return `naryResult` directly.  

If my understanding is correct, then there could potentially be an edge case, in which case the StaticType contains only `IntType`, and the result is somehow `Decimal`, which truly indicate an error in Type Inferencer. 

The slightly more complicated logic is then used to catch this scenario. 

Let me know if I am overcomplicating the problem here. If so, more than happy to simplify the logic.

Reply:
Ok, it wasn't clear to me of that edge case (i.e. when the inferencer expects just `IntType`s and the evaluation result isn't an Int) from the original PR. I think using the `integerValueValidator` to also check for the static type matching the evaluator behavior is confusing. Perhaps the following factoring is more clear:

```Kotlin
when (biggestIntegerType) {
    is IntType -> {
        val validator = integerValueValidator(biggestIntegerType.rangeConstraint.validRange)
        thunkFactory.thunkEnv(metas) { env ->
            val naryResult = computeThunk(env)
            when (naryResult.type) {
                ExprValueType.INT, ExprValueType.MISSING, ExprValueType.NULL  -> errorSignaler.errorIf(
                    !validator(naryResult),
                    ErrorCode.EVALUATOR_INTEGER_OVERFLOW,
                    { ErrorDetails(metas, "Integer overflow", errorContextFrom(metas)) },
                    { naryResult }
                )
                else -> {
                    if (staticTypes.all { it is IntType }) {
                        error("...") // <- here use the error that was previously part of `integerValueValidator`
                    } else {
                        naryResult
                    }
                }
            }
        }
    }
    // If there is no IntType StaticType, can't validate the integer size either.
    null -> computeThunk
    else -> computeThunk
}
```

Then for `integerValueValidator`, provide a different error such as saying the function only validates int, missing, and null types.